### PR TITLE
hacky solution for deadlock

### PR DESCRIPTION
**Pull request by acgreek**
not a great solution, please clean up

### CODE REVIEW

File: std/crypto/tls/conn.go [hunk 1]
```diff
@@ -1574,10 +1591,14 @@ func (c *Conn) release() {
 
 // Close closes the connection.
 func (c *Conn) Close() error {
-	c.closeMux.Lock()
+	if !c.hasLock {
```
Comment (lesismal) [inline std/crypto/tls/conn.go:1594]: here is still not a guard for concurrent operations:
1. another goroutine is locking, so we get a `c.hasLock == true` here and not call c.closeMux.Lock()

File: std/crypto/tls/conn.go [hunk 2]
```diff
@@ -1574,10 +1591,14 @@ func (c *Conn) release() {
...
+		c.closeMux.Lock()
+	}
 	closed := c.closed
 	c.closed = true
-	c.closeMux.Unlock()
+	if !c.hasLock {
```
Comment (lesismal) [inline std/crypto/tls/conn.go:1599]: 2. another goroutine set c.hasLock = false before we get c.hasLock again.
3. we get a `c.hasLock == false` and call c.closeMux.Unlock(), without the previos Lock(), then we went wrong.

File: std/crypto/tls/conn.go [hunk 3]
```diff
@@ -1390,8 +1395,12 @@ func (c *Conn) handleKeyUpdate(keyUpdate *keyUpdateMsg) error {
 
 // Append .
 func (c *Conn) Append(b []byte) (int, error) {
-	c.closeMux.Lock()
-	defer c.closeMux.Unlock()
+	if !c.hasLock {
```
Comment (acgreek) [inline std/crypto/tls/conn.go:1398]: This is another race condition, reading it outside the lock, then setting it inside the lock. probably should read it again inside the lock; but what would that mean if it has a value true after reading it as having a value of false?  having A recursive lock is probably the correct approach if this was c/c++ code.

File: std/crypto/tls/conn.go [hunk 4]
```diff
@@ -1390,8 +1395,12 @@ func (c *Conn) handleKeyUpdate(keyUpdate *keyUpdateMsg) error {
...
+		c.closeMux.Lock()
+		c.hasLock = true
+		defer c.closeMux.Unlock()
+		defer func() { c.hasLock = false }()
```
Comment (acgreek) [inline std/crypto/tls/conn.go:1402]:
```suggestion
		defer func() { 
		   c.hasLock = false 
		   c.closeMux.Unlock()
		}()
```

**Comment by acgreek** [comment]
Yeah, I know this locking code is confusing and bad. I would need to think
about it some more if I were to clean it up but it is too late in the day
for me; That is why I called it a hacky solution. If you have the cycles;
please come up with a better solution

**Comment by lesismal** [comment]
> Yeah, I know this locking code is confusing and bad. I would need to think about it some more if I were to clean it up but it is too late in the day for me; That is why I called it a hacky solution. If you have the cycles; please come up with a better solution

ok, good night, I'm thinking about it too

**Comment by lesismal** [comment]
It seems like this:
tls.Conn.Close() -> 
nbio.Conn.Close() -> 
nbio.Gopher.onCLose() ->
tls.Conn.Close() 
and we get the deadlock here.

Maybe I have get the solusion:
```golang
// OnClose registers callback for disconnected
func (g *Gopher) OnClose(h func(c *Conn, err error)) {
  if h == nil {
    panic("invalid nil handler")
  }
  g.onClose = func(c *Conn, err error) {
    g.afterFunc(1, func() {
      h(c, err)
    })
  }
}
```

Execute the nbio.Gopher.onClose in another goroutine, then we should solve it.

**Comment by lesismal** [comment]
already fixed in nbio.

**Pull request closed without merging**